src/main/java/com/aws/greengrass/clientdevices/auth/PermissionEvaluationUtils.java

```diff
@@ -56,118 +56,198 @@ public PermissionEvaluationUtils(GroupManager groupManager) {
+        }
+
+        String service = safeSubstring(resourceStr, 0, split);
+        if (service.isEmpty() || !StringUtils.isAlpha(service)) {
+            throw new PolicyException(EXCEPTION_MALFORMED_RESOURCE);
         }
 
-        Matcher matcher = SERVICE_RESOURCE_PATTERN.matcher(resourceStr);
-        if (matcher.matches()) {
-            return Resource.builder().resourceStr(resourceStr)
-                    .service(matcher.group(1))
-                    .resourceType(matcher.group(2))
-                    .resourceName(matcher.group(3))
-                    .build();
+        String typeAndName = safeSubstring(resourceStr, split + 1);
+        if (typeAndName.isEmpty()) {
+            throw new PolicyException(EXCEPTION_MALFORMED_RESOURCE);
         }
 
-        throw new IllegalArgumentException(
-                String.format("Resource %s is not in the form of %s", resourceStr, SERVICE_RESOURCE_PATTERN.pattern()));
+        split = typeAndName.indexOf(':');
+        if (split == -1) {
+            throw new PolicyException(EXCEPTION_MALFORMED_RESOURCE);
+        }
+
+        String resourceType = safeSubstring(typeAndName, 0, split);
+        if (resourceType.isEmpty() || !StringUtils.isAlpha(resourceType)) {
+            throw new PolicyException(EXCEPTION_MALFORMED_RESOURCE);
+        }
+
+        String resourceName = safeSubstring(typeAndName, split + 1);
+
+        // still using regex because Pattern.UNICODE_CHARACTER_CLASS is complicated
+        if (!RESOURCE_NAME_PATTERN.matcher(resourceName).matches()) {
+            throw new PolicyException(EXCEPTION_MALFORMED_RESOURCE);
+        }
+
+        return Resource.builder()
+                .resourceStr(resourceStr)
+                .service(service)
+                .resourceType(resourceType)
+                .resourceName(resourceName)
+                .build();
+    }
+
+    private static boolean isAlphanumericWithExtraChars(CharSequence cs, String extra) {
+        if (Utils.isEmpty(cs)) {
+            return false;
+        }
+        for (int i = 0; i < cs.length(); i++) {
+            char curr = cs.charAt(i);
+            if (!Character.isLetterOrDigit(curr) && extra.indexOf(curr) == -1) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    /**
+     * Like {@link String#substring(int, int)}, except rather than throwing, start/end
+     * indexes will be clamped to the string's length.
+     * <ul>
+     * <li>safeSubstring("a", 0, 0) -> ""</li>
+     * <li>safeSubstring("a", 0, 1) -> "a"</li>
+     * <li>safeSubstring("a", 0, 10) -> "a"</li>
+     * <li>safeSubstring("a", 1, 10) -> ""</li>
+     * </ul>
+     *
+     * @param str   input string
+     * @param start substring start, inclusive. must be greater than zero.
+     * @param end   substring end,
+     * @return substring
+     */
+    private static String safeSubstring(@NonNull String str, int start, int end) {
```

Comment:
nonnull will _throw_ if it is null. Is that what you want?

Reply:
Is this really any safer or just not doing what you expect it to do. This seems more confusing to me than helpful IMO

Reply:
Maybe just needs a rename: "substringWhenInBounds" for example

Reply:
if it's more confusing doesn't seem worth it, will just add checking inline

Reply:
done, also added more tests